homeassistant/components/esphome/__init__.py

```diff
@@ -312,6 +316,9 @@ async def complete_setup() -> None:
         await entry_data.async_update_static_infos(hass, entry, infos)
         await _setup_services(hass, entry_data, services)
 
+        if entry_data.device_info is not None and entry_data.device_info.name:
+            cli.expected_name = entry_data.device_info.name
```

Comment:
After I [commented](https://github.com/home-assistant/core/issues/52239#issuecomment-1019135096) in #52239 I realised the name should be available and we can set that on the `reconnect_logic`. It works without this PR, but could just add at the same time
```suggestion
        if entry_data.device_info is not None and entry_data.device_info.name:
            cli.expected_name = entry_data.device_info.name
            reconnect_logic.name = entry_data.device_info.name
```

Reply:
An yes that must have been lost somewhere in the ReconnectLogic migration. Good catch!